components/producers/blackduck/main.go

```diff
@@ -0,0 +1,204 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+	"strings"
+	"time"
+
+	v1 "github.com/ocurity/dracon/api/proto/v1"
+
+	"github.com/ocurity/dracon/components/producers"
+	packageurl "github.com/package-url/packageurl-go"
+)
+
+func main() {
+	if err := producers.ParseFlags(); err != nil {
+		log.Fatal(err)
+	}
+
+	inFile, err := producers.ReadInFile()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var results BlackduckOut
+	if err := producers.ParseJSON(inFile, &results); err != nil {
+		log.Fatal(err)
+	}
+
+	issues, err := parseIssues(&results)
+	if err != nil {
+		log.Fatal(err)
+	}
+	if err := producers.WriteDraconOut(
+		"gosec",
+		issues,
+	); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func cweIdsToInt(ids []string) []int32 {
+	res := []int32{}
+	for _, id := range ids {
+		newID := strings.ReplaceAll(id, "CWE-", "")
+		newIntID, err := strconv.Atoi(newID)
+		if err != nil {
+			log.Println(err)
+			continue
+		}
+		res = append(res, int32(newIntID))
+	}
+	return res
+}
+
+func BDTargetToPurl(originName, originID string) string {
+	splitOrigin := strings.Split(originID, ":")
+	return packageurl.NewPackageURL(originName, splitOrigin[0], splitOrigin[1], splitOrigin[2], packageurl.Qualifiers{}, "").ToString()
```

Comment:
Unfortunately the origin name doesn't always map 1:1 with the [recognized PURL types](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst). I am unable to locate BD docs about what origins it may report, but [this KB article](https://community.synopsys.com/s/article/What-is-an-Origin-and-Origin-ID-in-Blackduck) already tells us that NPM packages will be reported with origin `npmjs`, whereas the correct PURL type is `npm`.

Consider adding a sanity-check here that logs a warning or so when `originName` isn't found in either [`KnownTypes` or `CandidateTypes`](https://pkg.go.dev/github.com/package-url/packageurl-go#pkg-variables). That should allow you to discover irregularities and add mappings over time.

Reply:
Non-exhaustive list of known mappings:

| BD Origin Name | PURL |
|:------------|:--------------|
| `anaconda` | [`pkg:conda/...`](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#conda) |
| `centos` | [`pkg:rpm/centos/...`](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#rpm) |
| `debian` | [`pkg:deb/debian/...`](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#deb) |
| `fedora` | [`pkg:rpm/fedora/...`](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#rpm) | 
| `gradle` | [`pkg:maven/...`](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#maven) |
| `npmjs` | [`pkg:npm/...`](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#npm) |
| `opensuse` | [`pkg:rpm/opensuse/...`](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#rpm) |
| `packagist` | [`pkg:composer/...`](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#composer) |
| `redhat` | [`pkg:rpm/redhat/...`](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#rpm) |
| `unknown` | [`pkg:generic/...`](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#generic) |

Reply:
can you please take another look?